lib/StackAnalysis/StackAnalysis.cpp

```diff
@@ -324,6 +354,569 @@ void commitToModel(GeneratedCodeBasicInfo &GCBI,
+        }
+      }
+    }
+  }
+
+  // Elect a final stack offset.
+  const auto WinFSO = electFSO(MaybeReturns);
+
+  // Did we find a valid return instruction?
+  for (const auto &E : MaybeReturns) {
+    auto *CI = E.first;
+    if (E.second == *WinFSO) {
+      Result.insert({ E.first, FunctionEdgeTypeValue::Return });
+      unsigned NumArgs = CI->getNumArgOperands();
+      if (NumArgs > 3) {
+        for (unsigned Idx = 3; Idx < NumArgs; ++Idx) {
+          if (isa<ConstantInt>(CI->getArgOperand(Idx))) {
+            if (cast<ConstantInt>(CI->getArgOperand(Idx))->getZExtValue() == 0)
+              CalleeSavedRegs.insert(ABIRegisters[Idx - 3]);
+          }
+        }
+      }
+    } else {
+      Result.insert({ E.first, FunctionEdgeTypeValue::BrokenReturn });
+    }
+  }
+
+  // Neither a return nor a broken return?
+  for (CallBase *Call : callers(IndirectBranchInfoMarker)) {
+    auto *CI = cast<CallInst>(Call);
+    if (CI->getParent()->getParent() == F) {
+      if (isa<ConstantInt>(CI->getArgOperand(0)))
+        continue;
+
+      if (WinFSO.has_value() && (isa<ConstantInt>(CI->getArgOperand(1)))
+          && (cast<ConstantInt>(CI->getArgOperand(1))->getZExtValue()
+              == *WinFSO))
+        Result.insert({ CI, FunctionEdgeTypeValue::IndirectTailCall });
+      else
+        Result.insert({ CI, FunctionEdgeTypeValue::LongJmp });
+    }
+  }
+
+  bool FoundRet, FoundBrokenRet;
+  FoundRet = FoundBrokenRet = false;
+  for (const auto &E : Result) {
+    if (E.second == FunctionEdgeTypeValue::Return)
+      FoundRet = true;
+    if (E.second == FunctionEdgeTypeValue::BrokenReturn)
+      FoundBrokenRet = true;
+  }
+
+  FunctionTypeValue Type;
+  if (FoundRet) {
+    Type = FunctionTypeValue::Regular;
+  } else if (FoundBrokenRet) {
+    Type = FunctionTypeValue::Fake;
+  } else {
+    Type = FunctionTypeValue::NoReturn;
+  }
+
+  // Finally retrieve the clobbered registers and return them.
+  std::erase_if(ClobberedRegs, [&](const auto &E) {
+    return CalleeSavedRegs.find(E) != CalleeSavedRegs.end();
+  });
+
+  // Refine ABI analyses results.
+  auto *SP = GCBI->spReg();
+  auto It = ABIResults.Arguments.find(SP);
+  if (It != ABIResults.Arguments.end())
+    ABIResults.Arguments.erase(SP);
+  
+  for (const auto &[K, V] : ABIResults.CallSites) {
+    auto It = ABIResults.CallSites[K].Arguments.find(SP);
+    if (It != ABIResults.CallSites[K].Arguments.end())
+      ABIResults.CallSites[K].Arguments.erase(SP);
+  }
+
+  for (const auto &Reg : CalleeSavedRegs) {
+    auto It = ABIResults.Arguments.find(Reg);
+    if (It != ABIResults.Arguments.end())
+      ABIResults.Arguments.erase(Reg);
+  }
+
+  for (const auto &[K, V] : ABIResults.CallSites) {
+    for (auto &Reg : CalleeSavedRegs) {
+      auto It = ABIResults.CallSites[K].Arguments.find(Reg);
+      if (It != ABIResults.CallSites[K].Arguments.end())
+        ABIResults.CallSites[K].Arguments.erase(Reg);
+    }
+  }
```

Comment:
You're duplicating code here. Consider using a set composed by `CalleeSavedRegs` and `sp` and just do one run.